cpp/src/arrow/array/builder_binary.h

```diff
@@ -48,7 +48,10 @@ class ARROW_EXPORT BinaryBuilder : public ArrayBuilder {
   Status Append(const uint8_t* value, int32_t length) {
     ARROW_RETURN_NOT_OK(Reserve(1));
     ARROW_RETURN_NOT_OK(AppendNextOffset());
-    ARROW_RETURN_NOT_OK(value_data_builder_.Append(value, length));
+    // Safety check for UBSAN.
```

Comment:
Probably.  The never passing a nullptr is kind of the approach I've taken in other places (but they feel a little hacky).  Please let me know if those look OK, and I can take a similar approach with these.

Reply:
Did you run the builder benchmarks before and after the change? If there's no significant difference, then looks ok to me.

Reply:
I haven't had a chance yet.  Should get to this in the next few days.  I will ping you when I do.

Reply:
I tried running benchmarks, and I don't trust them on my box.  The biggest thing that seemed to change was stdev (without corresponding changes to the actual benchmark).  Are the benchmarks setup yet on Ursa's hardware, can we use those?  Otherwise I will go through and find all instances of where we pass through nullptr as a precaution (or we could turn off the check).

Reply:
CC @fsaintjacques is this something Archery is ready to help with?

Reply:
https://github.com/apache/arrow/pull/4285 once this is merged, we can call @ursabot to the rescue.

Reply:
@emkornfield stddev is irrelevant unless it becomes too large. If the min or average numbers didn't change significantly, then I'd say things are good.

Reply:
@ursabot benchmark

Reply:
@fsaintjacques I'm not seeing an update from Ursa bot should I try to squash/force push?

Reply:
Only works with global comments it seems.